Add Ohme integration

dan-r opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds a new integration for Ohme, a UK-based company that manufactures EV chargers (EVSE). I have been maintaining this integration as a [custom integration](https://github.com/dan-r/homeassistant-ohme) for the past year. Before submitting this I have moved the API logic to [ohmepy](https://github.com/dan-r/ohmepy).

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: home-assistant/home-assistant.io#36216

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

dan-r committed:
Initial contribution of Ohme

dan-r committed:
Comments and formatting

dan-r committed:
More formatting

dan-r committed:
Fix hassfest failures

dan-r committed:
Continue fixing hassfest failures

dan-r committed:
Bump ohmepy

dan-r committed:
Fix test linting

dan-r committed:
More pylint fixes

dan-r committed:
Even more lint fixes

dan-r committed:
Fix mypy failure

dan-r committed:
Fix test imports

MartinHjelmare commented on script/hassfest/quality_scale.py:
```diff
@@ -752,6 +752,7 @@ class Rule:
     "octoprint",
     "oem",
     "ohmconnect",
+    "ohme",
```
Please add a quality scale file instead. New integrations require Bronze tier:

https://developers.home-assistant.io/docs/core/integration-quality-scale/

https://developers.home-assistant.io/docs/core/integration-quality-scale/#keeping-track-of-the-implemented-rules

dan-r replied:
Thanks for the quick response! I've got it to a stage where I believe it meets bronze tier.

dan-r committed:
Migeate from hass.data to runtime_data

dan-r committed:
Fix binary sensor bug

dan-r committed:
Get config flow to 100% test coverage

dan-r committed:
Fix linting failures

joostlek commented:
https://github.com/home-assistant/core/pull/132574#pullrequestreview-2486670775

dan-r commented:
> [#132574 (review)](https://github.com/home-assistant/core/pull/132574#pullrequestreview-2486670775)

Ah - I misinterpreted the terminology and initially assumed platform was referring to an integration, not the types of entities I am providing. 

So to move this forward would it be preferable to initially contribute the binary sensors (for example) and add the rest of the platforms in a follow-up PR (or multiple)?

joostlek commented:
Looking at the code, it will require some refactors before it's ready for core and that's fine, we'll guide you with this! But keep in mind that if you now pick a huge platform (usually sensor is big) with 4 coordinators and stuff, that the review process will take significantly longer. So I would advice to start small. I would at least recommend a stateful entity, since that also requires the data fetching to be implemented rather than just a button platform which only calls the library.

dan-r committed:
Strip down to ine platform

dan-r commented:
Right I've taken out all platforms but binary sensor. I really appreciate the patience and guidance!

Since splitting the API library out, I could also do with some guidance as to what logic should be shifted over to that too. Especially as I do have some processing logic on API responses.

zweckj commented on homeassistant/components/ohme/__init__.py:
```diff
@@ -0,0 +1,112 @@
+    coordinators: list[DataUpdateCoordinator]
+
+
+async def async_update_listener(hass, entry):
+    """Handle options flow credentials update."""
+
+    # Reload this instance
+    await hass.config_entries.async_reload(entry.entry_id)
```
remove options flow for initial PR

zweckj commented on homeassistant/components/ohme/__init__.py:
```diff
@@ -0,0 +1,112 @@
+
+    client = OhmeApiClient(entry.data["email"], entry.data["password"])
+
+    entry.runtime_data = OhmeRuntimeData(client, [])
```
```suggestion
```
should be enough if we assign it after the initial update, since you will have coordinators in any case

zweckj commented on homeassistant/components/ohme/__init__.py:
```diff
@@ -0,0 +1,112 @@
+    # We can function without these so setup can continue
+    coordinators_optional = [OhmeAdvancedSettingsCoordinator]
+
+    for coordinator in coordinators:
+        # Catch failures if this is an 'optional' coordinator
+        try:
+            await coordinator.async_config_entry_first_refresh()
+        except ConfigEntryNotReady:
+            allow_failure = False
+            for optional in coordinators_optional:
+                allow_failure = (
+                    True if isinstance(coordinator, optional) else allow_failure
+                )
+
+            if allow_failure:
+                _LOGGER.error(
+                    "%s failed to setup. This coordinator is optional so the integration will still function, but please raise an issue if this persists",
+                    coordinator.__class__.__name__,
+                )
+            else:
+                raise
```
- I would first refresh the required oordinators so we can avoid this complex construct with "allow_failure"
- the question is then, why are the advanced coordinators allowed to fail during setup but not later? They will raise `UpdateFailed`, so maybe they shouldn't raise but only log everytime?

dan-r replied:
When I first wrote this as a custom integration the API was having some reliability issues. I'll take this allow_failure stuff out.

zweckj commented on homeassistant/components/ohme/__init__.py:
```diff
@@ -0,0 +1,112 @@
+    # Setup entities
+    await hass.config_entries.async_forward_entry_setups(entry, ENTITY_TYPES)
+
+    entry.async_on_unload(entry.add_update_listener(async_update_listener))
```
```suggestion
```
not in first PR

zweckj commented on homeassistant/components/ohme/__init__.py:
```diff
@@ -0,0 +1,112 @@
+    return await hass.config_entries.async_unload_platforms(entry, ENTITY_TYPES)
+
+
+async def async_migrate_entry(
+    hass: core.HomeAssistant, config_entry: ConfigEntry
+) -> bool:
+    """Migrate old entry."""
+    # Version number has gone backwards
+    if config_entry.version > CONFIG_VERSION:
+        _LOGGER.error("Backwards migration not possible. Please update the integration")
+        return False
+
+    # Version number has gone up
+    if config_entry.version < CONFIG_VERSION:
+        _LOGGER.debug("Migrating from version %s", config_entry.version)
+        new_data = config_entry.data
+
+        config_entry.version = CONFIG_VERSION
+        hass.config_entries.async_update_entry(config_entry, data=new_data)
+
+        _LOGGER.debug("Migration to version %s successful", config_entry.version)
+
+    return True
```
```suggestion
```
we are putting it in now, there's nothing to migrate atm

zweckj commented on homeassistant/components/ohme/binary_sensor.py:
```diff
@@ -0,0 +1,105 @@
+    async_add_entities(sensors, update_before_add=True)
+
+
+class ConnectedBinarySensor(OhmeEntity, BinarySensorEntity):
```
those can be simplified with `SensorEntityDescriptions` to a single class

zweckj commented on homeassistant/components/ohme/manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "domain": "ohme",
+  "name": "Ohme",
+  "codeowners": ["@dan-r"],
+  "config_flow": true,
```
remove empty keys

zweckj commented on homeassistant/components/ohme/config_flow.py:
```diff
@@ -0,0 +1,136 @@
+class OhmeConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Config flow."""
+
+    VERSION = CONFIG_VERSION
```
```suggestion
```
1 is the default

zweckj commented on homeassistant/components/ohme/config_flow.py:
```diff
@@ -0,0 +1,136 @@
+    ) -> ConfigFlowResult:
+        """First config step."""
+
+        errors = {}
```
typing

zweckj commented on homeassistant/components/ohme/config_flow.py:
```diff
@@ -0,0 +1,136 @@
+            step_id="init",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        "email", default=self._config_entry.data["email"]
+                    ): str,
+                    vol.Optional("password"): str,
+                    vol.Required(
+                        "never_session_specific",
+                        default=self._config_entry.options.get(
+                            "never_session_specific", False
+                        ),
+                    ): bool,
+                    vol.Required(
+                        "never_collapse_slots",
+                        default=self._config_entry.options.get(
+                            "never_collapse_slots", False
+                        ),
+                    ): bool,
+                    vol.Required(
+                        "interval_chargesessions",
+                        default=self._config_entry.options.get(
+                            "interval_chargesessions", DEFAULT_INTERVAL_CHARGESESSIONS
+                        ),
+                    ): vol.All(
+                        vol.Coerce(float),
+                        vol.Clamp(min=DEFAULT_INTERVAL_CHARGESESSIONS),
+                    ),
+                    vol.Required(
+                        "interval_advanced",
+                        default=self._config_entry.options.get(
+                            "interval_advanced", DEFAULT_INTERVAL_ADVANCED
+                        ),
+                    ): vol.All(
+                        vol.Coerce(float), vol.Clamp(min=DEFAULT_INTERVAL_ADVANCED)
+                    ),
+                }
+            ),
+            errors=errors,
+        )
```
```suggestion
```
not in the initial PR

zweckj commented on homeassistant/components/ohme/coordinator.py:
```diff
@@ -0,0 +1,64 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class OhmeChargeSessionsCoordinator(DataUpdateCoordinator):
```
```suggestion
class OhmeChargeSessionsCoordinator(DataUpdateCoordinator[returnType]):
```
this is missing the return type of `_client.async_get_charge_sessions()`

zweckj commented on homeassistant/components/ohme/const.py:
```diff
@@ -0,0 +1,11 @@
+
+DOMAIN = "ohme"
+CONFIG_VERSION = 1
+ENTITY_TYPES = ["binary_sensor"]
```
this is normally called `PLATFORMS` and uses the enum `Platform` for its values

zweckj commented on homeassistant/components/ohme/coordinator.py:
```diff
@@ -0,0 +1,64 @@
+            name="Ohme Charge Sessions",
+            update_interval=timedelta(
+                minutes=config_entry.options.get(
+                    "interval_chargesessions", DEFAULT_INTERVAL_CHARGESESSIONS
```
we don't allow setting the update interval through the Options. Since this is the case, I'm really thinking if it makes sense at all to use different coordinators, as you could just refresh the advanced part on every second "normal" refresh and it would also make the rest of the code easier.

dan-r replied:
I think this approach makes a lot of sense. With the custom integration version, having 4 coordinators and some entities needing to handle multiple is a pain!

zweckj commented on homeassistant/components/ohme/entity.py:
```diff
@@ -0,0 +1,38 @@
+from homeassistant.helpers.entity import Entity
+
+
+class OhmeEntity(Entity):
```
```suggestion
class OhmeEntity(Entity):
```
those usally inherit from `CoordinatorEntity`

zweckj commented on homeassistant/components/ohme/entity.py:
```diff
@@ -0,0 +1,38 @@
+            self.coordinator.async_add_listener(self._handle_coordinator_update, None)
+        )
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        self.async_write_ha_state()
```
and then this is default

zweckj commented on homeassistant/components/ohme/entity.py:
```diff
@@ -0,0 +1,38 @@
+
+        self._attr_device_info = client.get_device_info()
+
+    async def async_added_to_hass(self) -> None:
+        """When entity is added to hass."""
+        await super().async_added_to_hass()
+        self.async_on_remove(
+            self.coordinator.async_add_listener(self._handle_coordinator_update, None)
+        )
```
and this as well

zweckj commented on homeassistant/components/ohme/entity.py:
```diff
@@ -0,0 +1,38 @@
+
+    _attr_has_entity_name = True
+
+    def __init__(self, cooordinator, hass, client):
```
typing

zweckj commented on homeassistant/components/ohme/__init__.py:
```diff
@@ -0,0 +1,112 @@
+    coordinators: list[DataUpdateCoordinator]
+
+
+async def async_update_listener(hass, entry):
```
typing

zweckj commented on homeassistant/components/ohme/manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "codeowners": ["@dan-r"],
+  "config_flow": true,
+  "dependencies": [],
+  "documentation": "https://github.com/dan-r/HomeAssistant-Ohme",
```
needs to be on home-assistant.io

zweckj commented on homeassistant/components/ohme/entity.py:
```diff
@@ -0,0 +1,38 @@
+        self._last_updated = None
+        self._state = None
+
+        self._attr_device_info = client.get_device_info()
```
this needs to have `DeviceInfo` type, so better to construct it here

zweckj reviewed (changes requested):
thanks, some initial feedback, you are also missing typing (and return types) in general

dan-r committed:
Initial review changes

dan-r committed:
Add types, refactor coordinator, use BinarySensorEntityDescriptions

dan-r committed:
Cleanup and fix data types

dan-r commented:
Thank you for reviewing everything so quickly and thoroughly

I've resolved all those comments and added in parameter and return types everywhere. The integration code is looking **so** much simpler, so I definitely understand the rationale behind these changes. 

I've also updated the documentation to reflect the single platform.

zweckj commented on homeassistant/components/ohme/__init__.py:
```diff
@@ -0,0 +1,47 @@
+class OhmeRuntimeData:
+    """Store volatile data."""
+
+    client: OhmeApiClient
```
tbh you can just make the client property in the coordinator "public" and then you don't need to store it again

zweckj commented on homeassistant/components/ohme/__init__.py:
```diff
@@ -0,0 +1,47 @@
+async def async_setup_entry(hass: core.HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Ohme from a config entry."""
+
+    client = OhmeApiClient(entry.data["email"], entry.data["password"])
```
use the `CONF_` consts for accessing those

zweckj commented on homeassistant/components/ohme/__init__.py:
```diff
@@ -0,0 +1,47 @@
+    """Set up Ohme from a config entry."""
+
+    client = OhmeApiClient(entry.data["email"], entry.data["password"])
+    await client.async_create_session()
+    await client.async_update_device_info()
```
do these need to be error handled? I would move those to `_async_setup` (which is called as part of `async_config_entry_first_refresh`). Then we have all client related operations in the coordinator. You can then also move the `OhmeApiClient` init to the coordinator's init (you will have a `self.config_entry` available there after calling `super().init`)

zweckj commented on homeassistant/components/ohme/binary_sensor.py:
```diff
@@ -0,0 +1,90 @@
+            return False
+
+        handlers = {
+            "car_connected": lambda: bool(
+                self.coordinator.data.charge_sessions
+                and self.coordinator.data.charge_sessions["mode"] != "DISCONNECTED"
+            ),
+            "car_charging": lambda: bool(
+                self.coordinator.data.charge_sessions
+                and self.coordinator.data.charge_sessions.get("power")
+                and self.coordinator.data.charge_sessions["power"].get("watt", 0) > 0
+            ),
+            "pending_approval": lambda: bool(
+                self.coordinator.data.charge_sessions
+                and self.coordinator.data.charge_sessions["mode"] == "PENDING_APPROVAL"
+            ),
+            "charger_online": lambda: bool(
+                self.coordinator.data.advanced_settings
+                and self.coordinator.data.advanced_settings.get("online", False)
+            ),
```
those can be a `is_on_fn` as part of your `entity_description`. check e.g. tedee for examples how that works

zweckj commented on homeassistant/components/ohme/binary_sensor.py:
```diff
@@ -0,0 +1,90 @@
+    client = config_entry.runtime_data.client
+    coordinator = config_entry.runtime_data.coordinator
+
+    sensors = [
+        OhmeBinarySensor(coordinator, hass, client, description)
+        for description in BINARY_SENSOR_DESCRIPTIONS
+    ]
+
+    async_add_entities(sensors, update_before_add=True)
```
```suggestion

    async_add_entities(
       OhmeBinarySensor(coordinator, hass, client, description)
       for description in BINARY_SENSOR_DESCRIPTIONS
    )
```
you are using a coordinator, no need to update here first

zweckj commented on homeassistant/components/ohme/entity.py:
```diff
@@ -0,0 +1,38 @@
+
+        self.entity_description = entity_description
+        self._attr_translation_key = entity_description.key
+        self._attr_device_info = DeviceInfo(**client.get_device_info())
```
not a big fan of this tbh. I'd rather see here what the device is going to look like

dan-r replied:
The current approach is that the API library returns a dictionary with the same keys as needed for DeviceInfo. Would you just explicitly pull each property here or take a different approach?

zweckj replied:
Actually yes, I think so. Otherwise we'd need to check the library to see what the DeviceInfo looks like and if there are ever changes made to its definition we need to touch the library.

zweckj commented on homeassistant/components/ohme/strings.json:
```diff
@@ -0,0 +1,34 @@
+  "entity": {
+    "binary_sensor": {
+      "car_connected": {
+        "name": "Car Connected"
```
```suggestion
        "name": "Car connected"
```

zweckj commented on homeassistant/components/ohme/strings.json:
```diff
@@ -0,0 +1,34 @@
+        "name": "Car Connected"
+      },
+      "car_charging": {
+        "name": "Car Charging"
```
```suggestion
        "name": "Car charging"
```

zweckj commented on homeassistant/components/ohme/strings.json:
```diff
@@ -0,0 +1,34 @@
+        "name": "Car Charging"
+      },
+      "pending_approval": {
+        "name": "Pending Approval"
```
```suggestion
        "name": "Pending approval"
```

zweckj commented on homeassistant/components/ohme/binary_sensor.py:
```diff
@@ -0,0 +1,90 @@
+BINARY_SENSOR_DESCRIPTIONS = [
+    BinarySensorEntityDescription(
+        key="car_connected",
+        icon="mdi:ev-plug-type2",
```
move those to `icons.json`

zweckj commented on homeassistant/components/ohme/binary_sensor.py:
```diff
@@ -0,0 +1,90 @@
+_LOGGER = logging.getLogger(__name__)
+
+BINARY_SENSOR_DESCRIPTIONS = [
+    BinarySensorEntityDescription(
+        key="car_connected",
+        icon="mdi:ev-plug-type2",
+        device_class=BinarySensorDeviceClass.PLUG,
+    ),
+    BinarySensorEntityDescription(
+        key="car_charging",
+        icon="mdi:battery-charging-100",
+        device_class=BinarySensorDeviceClass.BATTERY_CHARGING,
+    ),
```
those three are probably mutually exclusive right? So wouldn't it make more sense to use a normal `sensor` of type `enum` instead?

dan-r replied:
Indeed they are and the enum approach seems to match with other EVSE integrations.

Where would you put the logic  to determine the enum value? In the external API library or somewhere in the integration?

zweckj replied:
Ideally in the library

zweckj commented on homeassistant/components/ohme/binary_sensor.py:
```diff
@@ -0,0 +1,90 @@
+        key="pending_approval",
+        icon="mdi:alert-decagram",
+    ),
+    BinarySensorEntityDescription(
+        key="charger_online",
+        icon="mdi:web",
+        device_class=BinarySensorDeviceClass.CONNECTIVITY,
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
```
I would rather bake this into the availbility of your base entity, as if your charger is offline probably all entities get unavailable anyway

zweckj commented on homeassistant/components/ohme/entity.py:
```diff
@@ -0,0 +1,38 @@
+    @property
+    def unique_id(self) -> str:
+        """Return unique ID of the entity."""
+        return f"{self._client.serial}_{self._attr_translation_key}"
```
use the _attr_ instead

zweckj commented on homeassistant/components/ohme/entity.py:
```diff
@@ -0,0 +1,38 @@
+        self._attr_translation_key = entity_description.key
+        self._attr_device_info = DeviceInfo(**client.get_device_info())
+
+        self._hass = hass
```
I don't think you need to pass or save this

zweckj commented on homeassistant/components/ohme/config_flow.py:
```diff
@@ -0,0 +1,38 @@
+        errors: dict[str, str] = {}
+
+        if user_input is not None:
+            await self.async_set_unique_id(user_input["email"])
```
`CONF_`  consts everywhere here as well

dan-r committed:
Move from binary sensors to an enum sensor

dan-r committed:
Move state logic to API library and cleanup tests

dan-r committed:
Add in a few other easy sensors

dan-r commented:
Alright so:
- I've made some changes to the API library to shift the logic to it
- No more binary sensors - all 3 status value consolidated into one enum sensor
- Snuck in a few other easy sensors

dan-r committed:
Fixed CI failures and add suggested unit to power sensor

zweckj commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,19 @@
+  dependency-transparency: done
+  docs-actions: done
+  docs-high-level-description: done
+  docs-installation-instructions: done
```
missing an `{% configuration_basic %}` block

zweckj commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,19 @@
+  config-flow-test-coverage: done
+  config-flow: done
+  dependency-transparency: done
+  docs-actions: done
```
exempt, no custom actions

zweckj commented on homeassistant/components/ohme/config_flow.py:
```diff
@@ -0,0 +1,40 @@
+        errors: dict[str, str] = {}
+
+        if user_input is not None:
+            await self.async_set_unique_id(user_input[CONF_EMAIL])
```
email is not a good unique id, use `self._async_abort_entries_match({CONF_EMAIL: user_input[CONF_EMAIL]})`

zweckj commented on tests/components/ohme/test_config_flow.py:
I'd still like a test that just succeeds without an error first and a test for already_configured

zweckj commented on tests/components/ohme/test_config_flow.py:
```diff
@@ -0,0 +1,41 @@
+
+
+@pytest.mark.usefixtures("mock_session")
+async def test_config_flow(hass: HomeAssistant) -> None:
```
add a `mock_setup_entry` to avoid actually setting up the entry in those tests

dan-r committed:
Resolve review comments and introduce snapshot testing

dan-r commented on homeassistant/components/ohme/strings.json:
```diff
@@ -0,0 +1,39 @@
+        }
+      },
+      "ct_current": {
+        "name": "CT current"
```
I could explain this in the documentation? I've gone with the same terminology as the vendor uses. Its the current transformer attached to the charger which goes around the main electricity supply to your house.

zweckj replied:
if that is something that is clear to a user of this device, it is fine for me as is

dan-r commented on homeassistant/components/ohme/entity.py:
```diff
@@ -0,0 +1,44 @@
+        client = coordinator.client
+        self._attr_unique_id = f"{client.serial}_{entity_description.key}"
+
+        device_info: dict[str, Any] = client.device_info
```
Its not currently typed, no

dan-r committed:
Review changes

zweckj commented on homeassistant/components/ohme/sensor.py:
```diff
@@ -0,0 +1,100 @@
+    async_add_entities(
+        OhmeSensor(coordinator, description)
+        for description in SENSOR_DESCRIPTIONS
+        if description.is_supported_fn(client)
```
So if I had two charger I'd need to make two accounts? If yes, set the integration type two device in the manifest.

dan-r replied:
That's correct, changed

dan-r committed:
Change integration type

zweckj commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,69 @@
+  unique-config-entry: done
+
+  # Silver
+  action-exceptions: todo
```
exempt atm as you have no actions

zweckj commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,69 @@
+  # Silver
+  action-exceptions: todo
+  config-entry-unloading: done
+  docs-configuration-parameters: todo
```
exempt, no options low

zweckj commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,69 @@
+  docs-installation-parameters: done
+  entity-unavailable: done
+  integration-owner: done
+  log-when-unavailable: todo
```
```suggestion
  log-when-unavailable: done
```
coordinator does that for you

zweckj commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,69 @@
+  docs-supported-functions: todo
+  docs-troubleshooting: todo
+  docs-use-cases: todo
+  dynamic-devices: todo
```
exempt, only one device per config entry

zweckj commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,69 @@
+  icon-translations: todo
+  reconfiguration-flow: todo
+  repair-issues: todo
+  stale-devices: todo
```
exempt, only one device per config entry

zweckj commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,69 @@
+  # Gold
+  devices: done
+  diagnostics: todo
+  discovery: todo
```
could those devices be discovered, e.g. by DHCP?

dan-r replied:
Currently all their devices use mobile SIMs for data comms, so not at the moment.

zweckj replied:
Then this and `discovery-update-info` are exempt. Cloud only devices cannot be discovered.

zweckj commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,69 @@
+  entity-disabled-by-default: todo
+  entity-translations: todo
+  exception-translations: todo
+  icon-translations: todo
```
```suggestion
  icon-translations: done
```

zweckj commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,69 @@
+  entity-category: todo
+  entity-device-class: todo
+  entity-disabled-by-default: todo
+  entity-translations: todo
```
```suggestion
  entity-translations: done
```

zweckj commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,69 @@
+  docs-use-cases: todo
+  dynamic-devices: todo
+  entity-category: todo
+  entity-device-class: todo
```
```suggestion
  entity-device-class: done
```

zweckj commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,69 @@
+  exception-translations: todo
+  icon-translations: todo
+  reconfiguration-flow: todo
+  repair-issues: todo
```
exempt, no repairs

zweckj commented on tests/components/ohme/conftest.py:
```diff
@@ -0,0 +1,94 @@
+            CONF_EMAIL: "test@example.com",
+            CONF_PASSWORD: "hunter2",
+        },
+        unique_id="test@example.com",
```
is not the unique id anymore
```suggestion
```

zweckj commented on tests/components/ohme/test_config_flow.py:
```diff
@@ -0,0 +1,83 @@
+    assert not result["errors"]
+
+    # Successful login
+    with patch("ohme.OhmeApiClient.async_login", return_value=True):
```
can we not just use the `mock_client` fixture here?

zweckj commented on tests/components/ohme/test_config_flow.py:
```diff
@@ -0,0 +1,83 @@
+    assert not result["errors"]
+
+    # Failed login
+    with patch("ohme.OhmeApiClient.async_login", return_value=False):
```
and here?

zweckj commented on tests/components/ohme/test_config_flow.py:
```diff
@@ -0,0 +1,83 @@
+    )
+    assert result["type"] is FlowResultType.FORM
+
+    result2 = await hass.config_entries.flow.async_configure(
```
minor thing, but you can just override result here

zweckj commented on tests/components/ohme/test_sensor.py:
bonus points if you add a test where the update fails and you show that makes your sensors go unavailable

dan-r replied:
I've got a test working that causes the coordinator to fail and asserting state is STATE_UNAVAILABLE works. Is there a good way to test that all sensors are showing as unavailable? A snapshot helper function perhaps? Couldn't find any obvious examples in other integrations.

```
async def test_coordinator_update_fails(
    hass: HomeAssistant,
    freezer: FrozenDateTimeFactory,
    mock_config_entry: MockConfigEntry,
    mock_client: MagicMock
) -> None:
    """Test the ohme coordinator update fails."""
    with patch("homeassistant.components.ohme.PLATFORMS", [Platform.SENSOR]):
        await setup_integration(hass, mock_config_entry)

    mock_client.async_get_charge_session.side_effect = ApiException
    freezer.tick(timedelta(seconds=60))
    async_fire_time_changed(hass)
    await hass.async_block_till_done(wait_background_tasks=True)

    state = hass.states.get("sensor.ohme_home_pro_energy")
    assert state.state == STATE_UNAVAILABLE

```

zweckj replied:
one sensor should be enough, as they all use the same class anyway

zweckj commented on homeassistant/components/ohme/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+
+            instance = OhmeApiClient(user_input[CONF_EMAIL], user_input[CONF_PASSWORD])
+            if not await instance.async_login():
+                errors["base"] = "auth_error"
```
```suggestion
                errors["base"] = "invalid_auth"
```
is what it's usually valled and there's also a default string available

zweckj commented on homeassistant/components/ohme/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+            self._async_abort_entries_match({CONF_EMAIL: user_input[CONF_EMAIL]})
+
+            instance = OhmeApiClient(user_input[CONF_EMAIL], user_input[CONF_PASSWORD])
+            if not await instance.async_login():
```
just checked the lib and this looks like it needs a `try:` here to me

dan-r replied:
I've updated the library so it throws AuthException for login failure now, and catching both ApiException and AuthException in the integration.

zweckj commented on homeassistant/components/ohme/coordinator.py:
```diff
@@ -0,0 +1,59 @@
+        self._alternative_iteration: bool = True
+
+    async def _async_setup(self) -> None:
+        if not await self.client.async_login():
+            raise ConfigEntryError("Unable to login to Ohme")
+
+        if not await self.client.async_update_device_info():
+            raise ConfigEntryNotReady("Unable to get Ohme device information")
```
and those two look like they can throw as well

dan-r committed:
Improved error handling, testing and updated quality scale

dan-r committed:
Update integrations.json

zweckj commented on homeassistant/components/ohme/__init__.py:
```diff
@@ -0,0 +1,29 @@
+"""The ohme integration."""
```
```suggestion
"""Set up ohme integration."""
```

zweckj commented on homeassistant/components/ohme/coordinator.py:
```diff
@@ -0,0 +1,65 @@
+type OhmeConfigEntry = ConfigEntry[OhmeCoordinator]
+
+
+class OhmeCoordinator(DataUpdateCoordinator[OhmeApiClient]):
```
```suggestion
class OhmeCoordinator(DataUpdateCoordinator[None]):
```

only noticed that now, but since you're storing the client in the coordinator anyway and that is storing the data, you do not need to return anything during the update

zweckj commented on homeassistant/components/ohme/sensor.py:
```diff
@@ -0,0 +1,100 @@
+    @property
+    def native_value(self):
+        """Return the state of the sensor."""
+        return self.entity_description.value_fn(self.coordinator.data)
```
```suggestion
        return self.entity_description.value_fn(self.coordinator.client)
```
would then change to that

zweckj commented on homeassistant/components/ohme/coordinator.py:
```diff
@@ -0,0 +1,65 @@
+                "An unexpected response was returned by the API"
+            ) from e
+
+    async def _async_update_data(self) -> OhmeApiClient:
+        """Fetch data from API endpoint."""
+        try:
+            await self.client.async_get_charge_session()
+
+            # Fetch on every other update
+            if self._alternative_iteration:
+                await self.client.async_get_advanced_settings()
+        except ApiException as e:
+            raise UpdateFailed("Error communicating with API") from e
+        else:
+            self._alternative_iteration = not self._alternative_iteration
+
+            return self.client
```
```suggestion
    async def _async_update_data(self) -> None:
        """Fetch data from API endpoint."""
        try:
            await self.client.async_get_charge_session()

            # Fetch on every other update
            if self._alternative_iteration:
                await self.client.async_get_advanced_settings()
        except ApiException as e:
            raise UpdateFailed("Error communicating with API") from e
        else:
            self._alternative_iteration = not self._alternative_iteration
```

and you don't need to return anything here then

zweckj commented on tests/components/ohme/test_config_flow.py:
```diff
@@ -0,0 +1,85 @@
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+
+
+async def test_config_flow_fail(
```
this is missing parametrization to also test the `ApiException` now

zweckj commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,90 @@
+  action-exceptions:
+    status: exempt
+    comment: |
+      This integration has no custom actions.
```
```suggestion
      This integration has no custom actions and read-only platform only.
```

zweckj reviewed (changes requested):
Looking pretty good, couple of small things

dan-r committed:
Update coordinator and parametrize config flow test

zweckj commented on homeassistant/components/ohme/__init__.py:
```diff
@@ -0,0 +1,29 @@
+from .const import PLATFORMS
+from .coordinator import OhmeConfigEntry, OhmeCoordinator
+
+_LOGGER = logging.getLogger(__name__)
```
```suggestion
```
not used here

zweckj commented on homeassistant/components/ohme/sensor.py:
```diff
@@ -0,0 +1,100 @@
+from .coordinator import OhmeConfigEntry
+from .entity import OhmeEntity
+
+_LOGGER = logging.getLogger(__name__)
```
```suggestion
```
nor here

zweckj commented on tests/components/ohme/test_config_flow.py:
```diff
@@ -0,0 +1,94 @@
+    test_exception: Exception,
+    expected_error: str,
+) -> None:
+    """Test config flow."""
```
```suggestion
    """Test config flow errors."""
```

zweckj commented on homeassistant/components/ohme/sensor.py:
```diff
@@ -0,0 +1,100 @@
+    config_entry: OhmeConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up sensors and configure coordinator."""
```
```suggestion
    """Set up sensors. """
```

zweckj commented on tests/components/ohme/conftest.py:
```diff
@@ -0,0 +1,105 @@
+        client.ct_connected = True
+        client.energy = 1000
+        client.device_info = {
+            "identifiers": ("ohme", "ohme_charger_chargerid"),
+            "name": "Ohme Home Pro",
+            "manufacturer": "Ohme",
+            "model": "Home Pro",
+            "sw_version": "v2.65",
+            "serial_number": "chargerid",
```
```suggestion
            "name": "Ohme Home Pro",
            "model": "Home Pro",
            "sw_version": "v2.65",
```
not a blocker, but I think those aren't used anymore in the integration

zweckj reviewed (changes requested):
only cleanups now from my end

dan-r committed:
Tidying up

zweckj commented on homeassistant/components/ohme/coordinator.py:
```diff
@@ -0,0 +1,63 @@
+        except AuthException as e:
+            raise ConfigEntryError("Unable to login to Ohme") from e
+        except ApiException as e:
+            raise ConfigEntryError(
```
```suggestion
            raise ConfigEntryNotReady(
```
missed that one. This should be correct as this error should probably be retried later.

dan-r committed:
Change coordinator exception type

zweckj commented on homeassistant/components/ohme/coordinator.py:
On second thought (and since you are so quick to respond 😉) let’s make those 3 exceptions translatable already (by setting `translation_domain` and `translation_key`, add them to strings and set the item done in the quality scale) because I think the users would profit

dan-r replied:
All done - I've put `PARALLEL_UPDATES = 0` into sensor.py too as it seems like an easy win with read only sensors and a coordinator.

dan-r committed:
Add exception translations

dan-r committed:
Whoops

zweckj reviewed (approved):
Thanks, LGTM! 👍

joostlek commented on homeassistant/components/ohme/coordinator.py:
```diff
@@ -0,0 +1,71 @@
+            config_entry=entry,
+        )
+
+        self.client: OhmeApiClient = OhmeApiClient(
```
```suggestion
        self.client = OhmeApiClient(
```
Can be inferred

joostlek commented on homeassistant/components/ohme/coordinator.py:
```diff
@@ -0,0 +1,71 @@
+            await self.client.async_login()
+
+            if not await self.client.async_update_device_info():
+                raise ConfigEntryNotReady(
```
```suggestion
                raise UpdateFailed(
```

joostlek commented on homeassistant/components/ohme/coordinator.py:
```diff
@@ -0,0 +1,71 @@
+                translation_key="auth_failed", translation_domain=DOMAIN
+            ) from e
+        except ApiException as e:
+            raise ConfigEntryNotReady(
```
```suggestion
            raise UpdateFailed(
```

dan-r replied:
Will UpdateFailed also cause the integration setup to fail? Because if this call fails and the entities get setup anyway, the device info will be all messed up

joostlek replied:
Yep, this is caught and rethrown as ConfigEntryNotReady

joostlek commented on homeassistant/components/ohme/coordinator.py:
```diff
@@ -0,0 +1,71 @@
+        try:
+            await self.client.async_get_charge_session()
+
+            # Fetch on every other update
+            if self._alternative_iteration:
+                await self.client.async_get_advanced_settings()
```
What's this for?

dan-r replied:
Some states (currently only CT reading) come from a different API endpoint. This is just to fetch that little less often, and save using a second coordinator

joostlek replied:
Your original code used to have 4, how will you add the other 2?

dan-r replied:
3 of those 4 need to be updated frequently (30s/1m), with the fourth being less frequent (10m). 

Aiming to keep things as simple as possible, my proposed solution in future PRs would be to make the 3 requests in OhmeCoordinator then add a second coordinator for the last.

joostlek replied:
I think i secretly like a separate coördinator for it. We for example have a service action to update the state of an entity, and that service now can't properly be used on one of the 1 minute entities as it will only work every other service action execution. And the other way around, if someone uses that service on the 30s one to request faster or slower, you automatically make the other request faster or slower as well. 

So like, yes I do see why you did this, but I think with this extra info I'm not sure this would make a good experience

dan-r replied:
So both coordinators are going to be calling methods of OhmeApiClient, and that is setup in OhmeCoordinator._async_setup. Whats the best way to have a second coordinator that depends on the actions of the first?

joostlek replied:
You would then execute those actions in `__init__.py` and then inject the client into the coordinator. (and then you won't implement `_async_setup`

joostlek commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,90 @@
+  # Gold
+  devices: done
+  diagnostics: todo
+  discovery:
+    status: exempt
+    comment: |
+      All supported devices are cloud connected. Discovery is not possible.
+  discovery-update-info:
+    status: exempt
+    comment: |
+      All supported devices are cloud connected. Discovery is not possible.
```
Did you check the following:
- Does the device emit mDNS packets?
- is the Mac address of the device a registered OUI?
- Does the device register with a certain hostname on the network?

dan-r replied:
Yep - all the devices they sell use mobile SIMs. No wifi or other LAN connection

joostlek commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,90 @@
+  docs-supported-functions: todo
+  docs-troubleshooting: todo
+  docs-use-cases: todo
+  dynamic-devices:
+    status: exempt
+    comment: |
+      This integration only provides one device per config entry.
```
No?

dan-r replied:
Should I set these two to todo? Out of interest how would you pass both with an integration that like this with one account, one device?

joostlek replied:
Oh i haven't read the other discussion. But you mean that you can only have one device in total? How did you come to that statement?

dan-r replied:
In the vendors app there is no option to link multiple chargers, only to unlink the existing charger or replace the current charger with another. As a result it's max one charger per account. As a residential product I kind of understand why they haven't bothered

joostlek replied:
How do their API look lik

dan-r replied:
It does return a list for chargeDevices, so I guess could support multiple devices in future, the API library just doesn't have any support for this yet

joostlek commented on homeassistant/components/ohme/quality_scale.yaml:
```diff
@@ -0,0 +1,90 @@
+    status: exempt
+    comment: |
+      This integration currently has no repairs.
+  stale-devices:
+    status: exempt
+    comment: |
+      This integration only provides one device per config entry.
```
No?

joostlek commented on homeassistant/components/ohme/strings.json:
```diff
@@ -0,0 +1,51 @@
+          "password": "[%key:common::config_flow::data::password%]"
+        },
+        "data_description": {
+          "email": "Enter the email address associated with your Ohme account. This should not be a social account (eg. Google).",
```
How do you mean should not be a social account?

dan-r replied:
As in if you signed in with Google when you created your account, you need to reset your password to use the integration. There's a note about this in the docs

joostlek replied:
Instead of explaining this in the data_description, I would just move the social part to the description of the step itself.

"This has to be an Ohme account. If you signed up with Ohme with a third party account like Google, please reset your password via Ohme, so you can create your own account"

I am probably cutting edges, but let's make it a little more descriptive. Bonus points if we can have a separate place on the integration docs where we explain this

joostlek commented on tests/components/ohme/conftest.py:
```diff
@@ -0,0 +1,102 @@
+        yield client
+
+
+@pytest.fixture(name="mock_session", autouse=True)
+def mock_session():
+    """Mock aiohttp.ClientSession."""
+    mocker = AiohttpClientMocker()
+
+    mocker.post(
+        "https://www.googleapis.com/identitytoolkit/v3/relyingparty/verifyPassword",
+        json={"idToken": "", "refreshToken": ""},
+    )
+
+    mocker.get(
+        "/v1/users/me/account", json=load_json_value_fixture("account.json", DOMAIN)
+    )
+
+    mocker.get(
+        "/v1/chargeSessions",
+        json=load_json_value_fixture("charge_sessions.json", DOMAIN),
+    )
+
+    mocker.get(
+        "/v1/chargeRules", json=load_json_value_fixture("charge_rules.json", DOMAIN)
+    )
+
+    mocker.get(
+        "/v1/chargeDevices/chargerid/advancedSettings",
+        json=load_json_value_fixture("advanced_settings.json", DOMAIN),
+    )
+
+    with patch(
+        "aiohttp.ClientSession",
+        side_effect=lambda *args, **kwargs: mocker.create_session(
+            asyncio.get_event_loop()
+        ),
+    ):
+        yield mocker
```
I would strongly recommend patching at the library instead of aiohttp

joostlek replied:
Patch the library, and have the mock return what the library returns. The big benefit of this is that we're only testing HA code (library code should be tested at the library) and you can change behaviour in the library without these tests being affected

joostlek commented on tests/components/ohme/conftest.py:
```diff
@@ -0,0 +1,102 @@
+        json=load_json_value_fixture("advanced_settings.json", DOMAIN),
+    )
+
+    with patch(
+        "aiohttp.ClientSession",
+        side_effect=lambda *args, **kwargs: mocker.create_session(
+            asyncio.get_event_loop()
+        ),
+    ):
```
Like we should definetly not patch this

joostlek commented on tests/components/ohme/test_config_flow.py:
```diff
@@ -0,0 +1,94 @@
+        result["flow_id"],
+        {CONF_EMAIL: "test@example.com", CONF_PASSWORD: "hunter2"},
+    )
+    await hass.async_block_till_done()
```
```suggestion
```

joostlek commented on tests/components/ohme/test_config_flow.py:
```diff
@@ -0,0 +1,94 @@
+        {CONF_EMAIL: "test@example.com", CONF_PASSWORD: "hunter2"},
+    )
+    await hass.async_block_till_done()
+    assert result["type"] is FlowResultType.CREATE_ENTRY
```
Let's also assert the data and title

joostlek commented on tests/components/ohme/test_config_flow.py:
```diff
@@ -0,0 +1,94 @@
+        result["flow_id"],
+        {CONF_EMAIL: "test@example.com", CONF_PASSWORD: "hunter1"},
+    )
+    await hass.async_block_till_done()
+    assert result["type"] is FlowResultType.FORM
+    assert result["errors"] == {"base": expected_error}
```
```suggestion
    assert result["type"] is FlowResultType.FORM
    assert result["errors"] == {"base": expected_error}
```
Let's have every config flow test end in either ABORT or CREATE_ENTRY

dan-r replied:
How would I end this test in an ABORT?

joostlek replied:
Not this test in particular, this one would be create entry, but once you've done that, all tests end in either one of those

joostlek commented on tests/components/ohme/test_config_flow.py:
```diff
@@ -0,0 +1,94 @@
+            CONF_PASSWORD: "hunter3",
+        },
+    )
+    await hass.async_block_till_done()
```
```suggestion
```

joostlek commented on tests/components/ohme/test_sensor.py:
```diff
@@ -0,0 +1,51 @@
+    mock_client: MagicMock,
+) -> None:
+    """Test that sensors show as unavailable after a coordinator failure."""
+    with patch("homeassistant.components.ohme.PLATFORMS", [Platform.SENSOR]):
+        await setup_integration(hass, mock_config_entry)
```
We don't have to limit the platforms here

dan-r replied:
Will it be needed when more platforms are added?

joostlek replied:
Nope

dan-r replied:
Ah - my mistake. Remove on line 39 but leave on line 26?

joostlek replied:
Yes, because it is needed for snapshot_platforms

joostlek commented on tests/components/ohme/test_sensor.py:
```diff
@@ -0,0 +1,51 @@
+    state = hass.states.get("sensor.ohme_home_pro_energy")
+    assert state.state == "1.0"
+
+    mock_client.async_get_charge_session.side_effect = ApiException
+    freezer.tick(timedelta(seconds=60))
+    async_fire_time_changed(hass)
+    await hass.async_block_till_done(wait_background_tasks=True)
```
If we test this, might as well also test that we are able to show a function if it works again

dan-r committed:
Review changes

dan-r committed:
More review changes

joostlek commented on homeassistant/components/ohme/coordinator.py:
```diff
@@ -3,69 +3,72 @@
 _LOGGER = logging.getLogger(__name__)
 
-type OhmeConfigEntry = ConfigEntry[OhmeCoordinator]
+type OhmeConfigEntry = ConfigEntry[list[DataUpdateCoordinator]]
```
I would consider making a dataclass for the coordinators, this way you can just access the individual coordinators without iterating the list

dan-r replied:
Got it - so then from sensors.py how would I pick out the correct coordinator?

joostlek replied:
Check withings, we do that there as well. 

```py
@dataclass(slots=True)
class WithingsData:
    """Dataclass to hold withings domain data."""

    client: WithingsClient
    measurement_coordinator: WithingsMeasurementDataUpdateCoordinator
    sleep_coordinator: WithingsSleepDataUpdateCoordinator
    bed_presence_coordinator: WithingsBedPresenceDataUpdateCoordinator
    goals_coordinator: WithingsGoalsDataUpdateCoordinator
```
So then I can access it with
`entry.runtime_data.measurement_coordinator`

joostlek commented on homeassistant/components/ohme/coordinator.py:
```diff
@@ -3,69 +3,72 @@
 
 
-class OhmeCoordinator(DataUpdateCoordinator[None]):
+class OhmeChargeSessionCoordinator(DataUpdateCoordinator[None]):
```
I would maybe recommend creating a base Coordinator. This way you can remove the constructor from the subcoordinators and you can automatically wrap all update statements in the same try catch so you don't have to duplicate it.

For example in Withings I have
```py


    async def _async_update_data(self) -> _DataT:
        try:
            return await self._internal_update_data()
        except (WithingsUnauthorizedError, WithingsAuthenticationFailedError) as exc:
            raise ConfigEntryAuthFailed from exc

    @abstractmethod
    async def _internal_update_data(self) -> _DataT:
        """Update coordinator data."""
```

(In your case you don't have different data types since both coordinators have the `None` type)

joostlek replied:
One of my coordinators then looks like this
```py

class WithingsDeviceDataUpdateCoordinator(
    WithingsDataUpdateCoordinator[dict[str, Device]]
):
    """Withings device coordinator."""

    coordinator_name: str = "device"
    _default_update_interval = timedelta(hours=1)

    async def _internal_update_data(self) -> dict[str, Device]:
        """Update coordinator data."""
        devices = await self._client.get_devices()
        return {device.device_id: device for device in devices}
```
Which is quite small and neat and without a lot of code duplication

dan-r replied:
You beat me to it :) Just started to do the same thing. That _internal_update_data idea saves a bit more duplication so I'll put that in too

dan-r committed:
Fix some data types

dan-r committed:
Move to a dataclass for runtime_data

joostlek commented on homeassistant/components/ohme/sensor.py:
```diff
@@ -0,0 +1,103 @@
+class OhmeSensorDescription(SensorEntityDescription):
+    """Class describing Ohme sensor entities."""
+
+    value_fn: Callable[[OhmeApiClient], Any]
```
Can we be more specific with `Any`?

joostlek commented on homeassistant/components/ohme/sensor.py:
```diff
@@ -0,0 +1,103 @@
+
+    value_fn: Callable[[OhmeApiClient], Any]
+    is_supported_fn: Callable[[OhmeApiClient], bool] = lambda _: True
+    coordinator_fn: Callable[[OhmeRuntimeData], OhmeBaseCoordinator] = (
+        lambda data: data.charge_session_coordinator
+    )
```
If you will add more coordinators, I think it would be more beautiful if we split the lists up and just inject the right one

joostlek commented on homeassistant/components/ohme/sensor.py:
```diff
@@ -0,0 +1,103 @@
+    """Generic sensor for Ohme."""
+
+    @property
+    def native_value(self):
```
return type

dan-r committed:
Review changes

joostlek commented on homeassistant/components/ohme/sensor.py:
```diff
@@ -0,0 +1,107 @@
+        (SENSOR_ADVANCED_SETTINGS, coordinators.advanced_settings_coordinator),
+    ]
+
+    for entities, coordinator in coordinator_map:
+        async_add_entities(
+            OhmeSensor(coordinator, description)
+            for description in entities
+            if description.is_supported_fn(coordinator.client)
+        )
```
```suggestion
    async_add_entities(
        OhmeSensor(coordinator, description)
        for entities, coordinator in coordinator_map
        for description in entities
        if description.is_supported_fn(coordinator.client)
    )
```
We should call `async_add_entities` ideally once, since its expensive

dan-r committed:
Only call async_add_entities once

joostlek commented:
@dan-r If you use discord, can you please send me a message?
